python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+        return deleted_tester_set
+
+    # Loop through the testers and add them to the runner pool for execution
+    def schedule(self, testers):
+        # We didn't recieve anything... pesky TestHarness
+        if testers == []:
+            return
+
+        # Immediately test for race conditions and return if condition exists
+        if self.checkRaceCondition(testers):
+            return
+
+        # Local storage of skipped tester objects
+        skipped_or_failed_tests = set([])
+
+        # Local storage of downstream deleted tester objects
+        downstream_deleted = set([])
+
+        # Local DAG
+        tester_dag = dag.DAG()
+
+        # Create the DAG, skipped_testers, and just about anything we can during this iteration loop
+        for tester in testers:
+            with self.thread_lock:
+                # Create tester name to object map
+                self.scheduled_name_to_object[tester.getTestName()] = tester
+
+            # Create a node for this tester if not already exists (prereq iteration creates nodes)
+            if not tester_dag.node_exists(tester.getTestName()):
+                tester_dag.add_node(tester.getTestName())
+
+            # Loop through prereqs and add leaves/nodes where necessary
+            for prereq in tester.getPrereqs():
+
+                # Discover unknown prereqs
+                if prereq not in [x.getTestName() for x in testers]:
+                    tester.setStatus('unknown dependency', tester.bucket_fail)
+                    skipped_or_failed_tests.add(tester)
+                    continue
+
+                if not tester_dag.node_exists(prereq):
+                    # Create a node for this tester
+                    tester_dag.add_node(prereq)
+
+                # Once you add an edge to a node, the DAG arranges the nodes accordingly
+                try:
+                    tester_dag.add_edge(prereq, tester.getTestName())
+
+                # Handle Cyclic errors (tester status has already be applied)
+                except dag.DAGValidationError:
+                    pass
+
+            # Discover skipped tests
+            if not tester.getRunnable(self.options):
+                skipped_or_failed_tests.add(tester)
+
+        # Discover skipped tests and remove downstream tests that require this skipped test
+        for tester in skipped_or_failed_tests:
+            # Delete and return downstream nodes (prints 'skipped dependency' tests to the screen)
+            downstream_deleted = self.deleteDownstreamTests(tester_dag, tester)
+
+            # Delete this test
+            tester_dag.delete_node_if_exists(tester.getTestName())
+
+            # Inform the user it was skipped by adding it to the queue
+            self.statusGo(tester.getTestName())
+
+        # Get top level jobs we can launch
+        ind_jobs = tester_dag.ind_nodes()
+
+        # Maybe all jobs in this group got skipped?
+        if ind_jobs == []:
```

Comment:
Careful here. While this effectively works it's logically incorrect. You know that because there are no independent nodes, there aren't any nodes. The only other case would be a cyclic graph which you've already handled. Logically it would be better to short circuit if the graph is empty instead.